routing/missioncontrol.go

```diff
@@ -226,100 +245,37 @@ func (m *MissionControl) GetProbability(fromNode, toNode route.Vertex,
-func (m *MissionControl) getProbAfterFail(lastFailure time.Time) float64 {
-	if lastFailure.IsZero() {
-		return m.cfg.AprioriHopProbability
-	}
-
-	timeSinceLastFailure := m.now().Sub(lastFailure)
-
-	// Calculate success probability based on the weight of the last
-	// failure. When the failure is fresh, its weight is 1 and we'll return
-	// probability 0. Over time the probability recovers to the a priori
-	// probability.
-	weight := m.getWeight(timeSinceLastFailure)
-	probability := m.cfg.AprioriHopProbability * (1 - weight)
-
-	return probability
-}
-
-// getWeight calculates a weight in the range [0, 1] that should be assigned to
-// a payment result. Weight follows an exponential curve that starts at 1 when
-// the result is fresh and asymptotically approaches zero over time. The rate at
-// which this happens is controlled by the penaltyHalfLife parameter.
-func (m *MissionControl) getWeight(age time.Duration) float64 {
-	exp := -age.Hours() / m.cfg.PenaltyHalfLife.Hours()
-	return math.Pow(2, exp)
-}
-
-// getLastPairResult gets the last recorded result for a node pair.
-func (m *MissionControl) getLastPairResult(fromNode,
-	toNode route.Vertex) *timedPairResult {
+	now := m.now()
+	results := m.lastPairResult[fromNode]
 
-	nodePairs, ok := m.lastPairResult[fromNode]
-	if !ok {
-		return nil
-	}
-
-	lastResult, ok := nodePairs[toNode]
-	if !ok {
-		return nil
-	}
-
-	return &lastResult
+	return m.estimator.getPairProbability(now, results, toNode, amt)
 }
 
 // setLastPairResult stores a result for a node pair.
 func (m *MissionControl) setLastPairResult(fromNode,
-	toNode route.Vertex, result *timedPairResult) {
+	toNode route.Vertex, result timedPairResult) {
 
 	nodePairs, ok := m.lastPairResult[fromNode]
 	if !ok {
 		nodePairs = make(NodeResults)
 		m.lastPairResult[fromNode] = nodePairs
 	}
 
-	nodePairs[toNode] = *result
+	nodePairs[toNode] = result
 }
 
-// getPairProbability estimates the probability of successfully
-// traversing from fromNode to toNode based on historical payment outcomes.
-func (m *MissionControl) getPairProbability(fromNode,
-	toNode route.Vertex, amt lnwire.MilliSatoshi) float64 {
-
-	// Start by getting the last node level failure. A node failure is
-	// considered a failure that would have affected every edge. Therefore
-	// we insert a node level failure into the history of every channel. If
-	// there is none, lastFail will be zero.
-	lastFail := m.lastNodeFailure[fromNode]
-
-	// Retrieve the last pair outcome.
-	lastPairResult := m.getLastPairResult(fromNode, toNode)
-
-	// Only look at the last pair outcome if it happened after the last node
-	// level failure. Otherwise the node level failure is the most recent
-	// and used as the basis for calculation of the probability.
-	if lastPairResult != nil && lastPairResult.timestamp.After(lastFail) {
-		if lastPairResult.success {
-			return prevSuccessProbability
-		}
+// setAllPairResult stores a result for all known connection of the given node.
```

Comment:
should be `setAllFail`, since only used for failure?

Reply:
It describes what the function does

Reply:
I think an earlier version of the PR changed this to `setAllFail`, since we never use it to set successes?

Reply:
I don't remember exactly, but I think in that version the function could also only set a failure.

Reply:
question is why we need one that can set success if that is never used

Reply:
yes, yagni, fixed.